ghost/milestone-emails/lib/MilestonesEmailService.js

```diff
@@ -0,0 +1,149 @@
+    }
+
+    /**
+     *
+     * @param {Array} goalValues
+     * @param {number} current
+     *
+     * @returns {Array}
+     */
+    _getMatchedMilestone = (goalValues, current) => {
+        // return highest suitable milestone
+        return goalValues.filter(value => current >= value)
+            .sort((a, b) => b - a)[0];
+    };
+
+    /**
+     *
+     * @param {number} milestone
+     * @param {'arr'|'members'} type
+     * @param {boolean} hasMembersImported
+     *
+     * @returns {Promise<Milestone>}
+     */
+    _saveMileStoneAndSendEmail = async (milestone, type, hasMembersImported) => {
+        const milestoneData = {
+            type,
+            value: milestone
+        };
+
+        if (type === 'arr') {
+            milestoneData.currency = this.#defaultCurrency;
+        }
+
+        // Two cases in which we don't want to send an email
+        // 1. There has been an import of members within the last week
+        // 2. The last email has been sent less than two weeks ago
+        const shouldSendEmail = await this.#api.shouldSendEmail();
+
+        if (shouldSendEmail && !hasMembersImported) {
+            // TODO: hook up GhostMailer or use StaffService and trigger event to send email
+            // await this.#mailer.send({
+            //     subject: 'Test',
+            //     html: '<div>Milestone achieved</div>',
+            //     to: 'test@example.com'
+            // });
+
+            milestoneData.emailSentAt = new Date();
+        }
+
+        const savedMilestone = await this.#api.checkAndProcessMilestone(milestoneData);
+
+        return savedMilestone;
+    };
+
+    async runARRQueries() {
+        // Fetch the current data from queries
+        const currentARR = await this.#queries.getARR();
+        const hasMembersImported = await this.#queries.hasImportedMembersInPeriod();
+
+        // Check the definitions in the config
+        const arrMilestoneSettings = this.#config.milestones.arr;
+
+        // First check the currency matches
+        if (currentARR.length) {
+            let milestone;
+
+            const currentARRForCurrency = currentARR.filter(arr => arr.currency === this.#defaultCurrency)[0];
+            const milestonesForCurrency = arrMilestoneSettings.filter(milestoneSetting => milestoneSetting.currency === this.#defaultCurrency)[0];
+
+            if (milestonesForCurrency && currentARRForCurrency) {
+                // get the closest milestone we're over now
+                milestone = this._getMatchedMilestone(milestonesForCurrency.values, currentARRForCurrency.arr);
+
+                // Fetch the latest milestone for this currency
+                const latestMilestone = await this.#api.getLatestArrMilestone(this.#defaultCurrency);
+
+                if (!latestMilestone || milestone > latestMilestone.value) {
+                    return await this._saveMileStoneAndSendEmail(milestone, 'arr', hasMembersImported);
```

Comment:
Not sure what what we should expect to be returned from the service. For now I returned the created milestone if there was one created.